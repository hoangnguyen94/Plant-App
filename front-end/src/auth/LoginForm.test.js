import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { useHistory, BrowserRouter } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import LoginForm from './LoginForm';
import { act } from 'react-dom';

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'), // Keep all the actual named exports
	useHistory: jest.fn(),
  }));

describe('LoginForm component', () => {
  beforeEach(() => {
    useHistory.mockReturnValue({
      push: jest.fn(),
    });
  });

	test( 'renders the login form component', () =>
	{
		const { container } = render( <LoginForm login={() => { }} /> );
    
		expect( container.firstChild ).toMatchSnapshot();
	} );

	test( 'handles form submission successfully', async () =>
	{
		const history = createMemoryHistory();
		const loginMock = jest.fn( () => Promise.resolve( { success: true } ) );
		const { getByLabelText, getByText } = render( <LoginForm login={loginMock} />, { history } );
  
		const usernameInput = getByLabelText( 'Username' );
		const passwordInput = getByLabelText( 'Password' );
		const submitButton = getByText( 'Submit' );
  
		fireEvent.change( usernameInput, { target: { value: 'testuser' } } );
		fireEvent.change( passwordInput, { target: { value: 'testpassword' } } );
		fireEvent.click( submitButton );
  
		expect( loginMock ).toHaveBeenCalledWith( {
			username: 'testuser',
			password: 'testpassword',
		} );
		expect( history.location.pathname ).toBe( '/' );
	} );
  
	test( 'handles form submission with errors', async () =>
	{
		const history = createMemoryHistory();
		const loginMock = jest.fn( () =>
			Promise.resolve( { success: false, errors: [ 'Invalid username', 'Invalid password' ] } )
		);
	
		const { getByLabelText, getByText, getByRole } = render(
			<BrowserRouter>
				<LoginForm login={loginMock} />
			</BrowserRouter>
		);
	
		const usernameInput = getByLabelText( 'Username' );
		const passwordInput = getByLabelText( 'Password' );
		const submitButton = getByText( 'Submit' );
	
		fireEvent.change( usernameInput, { target: { value: 'testuser' } } );
		fireEvent.change( passwordInput, { target: { value: 'testpassword' } } );
		fireEvent.click( submitButton );
	
		expect( loginMock ).toHaveBeenCalledWith( {
			username: 'testuser',
			password: 'testpassword',
		} );
		expect( history.location.pathname ).toBe( '/' );
	} );
});

