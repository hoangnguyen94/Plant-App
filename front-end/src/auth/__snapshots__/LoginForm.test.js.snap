// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component renders the login form component 1`] = `
<div
  class="LoginForm"
>
  <div
    class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
  >
    <h3
      class="mb-3"
    >
      Log In
    </h3>
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <form>
          <div
            class="form-group"
          >
            <label
              for="username"
            >
              Username
            </label>
            <input
              autocomplete="username"
              class="form-control"
              id="username"
              name="username"
              required=""
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="password"
            >
              Password
            </label>
            <input
              autocomplete="current-password"
              class="form-control"
              id="password"
              name="password"
              required=""
              type="password"
              value=""
            />
          </div>
          <button
            class="btn btn-primary float-right"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
